v3/integration/genTestCerts/genTestCerts.go

```diff
@@ -0,0 +1,314 @@
+		PermittedIPRanges:           nil,
+		ExcludedIPRanges:            nil,
+		PermittedEmailAddresses:     nil,
+		ExcludedEmailAddresses:      nil,
+		PermittedURIDomains:         nil,
+		ExcludedURIDomains:          nil,
+		CRLDistributionPoints:       nil,
+		PolicyIdentifiers:           nil,
+	}
+	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	cert, err := x509.CreateCertificate(rand.Reader, &template, parent, key.Public(), key)
+	if err != nil {
+		return nil, err
+	}
+	return x509.ParseCertificate(cert)
+}
+
+// Formats the given certificate into OpenSSL's textual output. For example:
+//
+// Certificate:
+//    Data:
+//        Version: 3 (0x2)
+//        Serial Number: 1 (0x1)
+//        Signature Algorithm: ecdsa-with-SHA256
+//        Issuer:
+//        Validity
+//            Not Before: Feb 14 17:21:17 2021 GMT
+//            Not After : Feb 14 17:21:17 2021 GMT
+//        Subject:
+//        Subject Public Key Info:
+//            Public Key Algorithm: id-ecPublicKey
+//                Public-Key: (256 bit)
+//                pub:
+//                    04:76:2b:19:b8:b4:f4:d9:9e:66:8a:6a:f3:bf:c5:
+//                    df:83:43:d6:53:bf:9e:5a:b8:b1:5d:99:8c:4e:d7:
+//                    59:25:fd:5c:08:16:23:19:61:c4:cc:c2:f7:db:ac:
+//                    72:a5:5e:65:35:f3:64:e2:9b:af:f9:04:c9:99:61:
+//                    57:3e:ee:9c:b3
+//                ASN1 OID: prime256v1
+//                NIST CURVE: P-256
+//        X509v3 extensions:
+//            X509v3 Subject Key Identifier:
+//                6E:3F:50:3A:07:4E:10:AA:74:31:8F:3B:B3:4F:30:96:D3:6F:EF:AE
+//    Signature Algorithm: ecdsa-with-SHA256
+//         30:44:02:20:11:3f:4a:25:63:10:fa:2d:96:00:e8:23:8c:62:
+//         40:c4:8d:31:31:d0:96:f2:7d:28:34:3a:2c:23:9f:bb:28:7e:
+//         02:20:1b:8a:68:6d:ef:c4:d7:19:46:48:bf:b0:18:85:31:37:
+//         ce:2f:04:27:7c:a3:d2:47:4d:e1:1f:c3:1a:3e:e3:8f
+// -----BEGIN CERTIFICATE-----
+// MIIBDjCBtqADAgECAgEBMAoGCCqGSM49BAMCMAAwHhcNMjEwMjE0MTcyMTE3WhcN
+// MjEwMjE0MTcyMTE3WjAAMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEdisZuLT0
+// 2Z5mimrzv8Xfg0PWU7+eWrixXZmMTtdZJf1cCBYjGWHEzML326xypV5lNfNk4puv
+// +QTJmWFXPu6cs6MhMB8wHQYDVR0OBBYEFG4/UDoHThCqdDGPO7NPMJbTb++uMAoG
+// CCqGSM49BAMCA0cAMEQCIBE/SiVjEPotlgDoI4xiQMSNMTHQlvJ9KDQ6LCOfuyh+
+// AiAbimht78TXGUZIv7AYhTE3zi8EJ3yj0kdN4R/DGj7jjw==
+// -----END CERTIFICATE-----
+//
+// Requires a copy of openssl in $PATH as it is simply making a
+// subprocess call out to it.
+func OpenSSLFormatCertificate(cert *x509.Certificate) (string, error) {
+	block := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert.Raw,
+	})
+	cmd := exec.Command("openssl", "x509", "-text")
+	cmd.Stdin = strings.NewReader(string(block))
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+	return string(output), nil
+}
+
+// Just a simple, sequential, serial number generator.
+var NextSerial = func() func() *big.Int {
```

Comment:
This is neat but a random serial number probably would have worked OK too :laughing: If the target audience of this script is folks not very familiar with Go a generator function that uses a sync primitive is probably going to seem somewhat magical. It's more robust, but this is a test script and so I think simplicity might be a better tradeoff.

Reply:
I think I'll have both - one sequential and one random (because lord knows at someone point we're going to need to lint something to do with serial numbers with various properties).